[Jweng88] iP

Jweng88 opened the pull request:
# Bob

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Bob frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/Jweng88/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Jweng88 committed:
Rename, Greet, Exit

Jweng88 committed:
Echo

Jweng88 committed:
Add, List

Jweng88 committed:
Mark as Done

Jweng88 committed:
ToDos, Events, Deadlines

Jweng88 committed:
Automated Text UI Testing

Jweng88 committed:
Handle Errors

Jweng88 committed:
Delete

Jweng88 committed:
Save tasks and load data from hard disk

The code now allows:
* saving of tasks in hard disk automatically
* loading of data from hard disk when chatbot starts up

Jweng88 committed:
Teach the chatbot to understand dates and times

The code will allow the chatbot to understand:

* store deadline dates and times as a java.time.LocalDateTime in the task object
* store start and end of event dates and times as java.time.LocalDateTime in the task object

Jweng88 committed:
Make the code more OOP

The code has been refactored to extract out closely related code as

classes which includes:

* Abstract class: Command
* Classes extending Command class: AddDeadlineCommand, AddEventCommand,

AddTodoCommand, DeleteCommand, ExitCommand, ListCommand,

MarkAsDoneCommand, MarkAsUndoneCommand

* Classes: Ui, Storage, Parser, TaskList, Duke

Jweng88 committed:
Organise the classes into suitable java packages

The code has been organised into suitable java packages as followed:

* An overall duke package
* Packages in duke: command, exception, storage, task, ui

Jweng88 committed:
Merge branch 'add-gradle-support' into branch-Gradle

Jweng88 committed:
Add JUnit tests to test behaviour of the code

The code has two added test classes:

* TaskListTest class -- tests for addTask, deleteTask, get,
getting and deleting invalid task

* ToDoTest class -- tests for toString, markAsDone, markAsNotDone,

createTaskFromFormattedString

Jweng88 committed:
Package the App as a JAR file

Edited on the build.gradle file to successfully package the app
to an executable JAR file.

Jweng88 committed:
Add JavaDoc comments to the code

Added header comments to all non-private classes/methods,
and non-trivial private methods.

Jweng88 committed:
Tweak the code to comply with the coding standard

The code now complies with the given coding standard.

Jweng88 committed:
Add JavaDoc comments to the code

Added additional header comments to all class-level members.

Jweng88 committed:
Allows users to find tasks with a keyword

The code now allows the user to find a task by searching for a keyword.

Jweng88 committed:
Merge branch 'branch-A-CodingStandard'

Jweng88 committed:
Merge branch 'branch-Level-9'

Tim-Siu commented on src/main/java/duke/Duke.java:
Very clear modularization. Look good to me!🤗

Tim-Siu commented on src/main/java/duke/command/Parser.java:
Clear and readable string parsing! Great use of enums.

Tim-Siu commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,143 @@
+     * @param errorMessage The error message to be displayed.
+     */
+    public void showError(String errorMessage) {
+        System.out.println("☹ " + errorMessage);
```
I like the inclusion of a unicode face. I wonder if that will cause a potential compatibility issue, though.

Tim-Siu commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,37 @@
+     * @return A formatted string representation of the ToDo object.
+     */
+    @Override
+    public String toFormattedString() {
```
The naming of methods is all of the course standards. Good job!

Tim-Siu commented on src/main/java/duke/command/AddDeadlineCommand.java:
Comments for methods and class are very clear

Tim-Siu reviewed (commented):
LGTM! No violation of coding standards can be found.

Jweng88 committed:
Add a GUI to the chatbot

The code uses the JavaFX technology to implement the GUI.

Users can now interact with the chatbot using the GUI
rather than the text-based UI.

Jweng88 committed:
Use checkStyle to detect coding style violations

The code now follows the coding style guidelines.

Jweng88 committed:
Use checkStyle to detect coding style violations

In the previous commit,
the code had some violations in the coding style.

The code now follows coding style  guidelines
while adding GUI to the chatbot.

Jweng88 committed:
Merge branch 'branch-Level-10'

wesho1107 commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,24 @@
+
+import javafx.application.Application;
+
+/**
+ * The Launcher class serves as the entry point for the Duke application.
+ * It uses JavaFX's Application.launch method to launch the application by
+ * starting the Main class.
+ * This class is primarily used to work around classpath issues that can
+ * occur when launching JavaFX applications.
+ */
```
I am not sure if I like how lengthy the JavaDoc is. Perhaps you could make it more concise to 1-2 sentences.

wesho1107 reviewed (commented):
Excellent job! There are really no nits from your code. Good layout, good naming conventions, good JavaDoc implementation, good import organisation. It was very clear when reading your code.

For some classes, your JavaDoc may be lengthy (up to 3 lines), if possible, try to have a shorter summary.

Jweng88 committed:
Adding assertions to UI elements and task index

The code has some assumptions about the UI elements
being present and task indexes being positive at runtime.

However this might not be the case, so missing UI elements and
negative task indexes can lead to unexpected behaviours.

Adding assertions can help us to catch these missing UI elements
and negative indexes and stop the execution.

Using assertions is preferable as it is easy to implement, easy to
disable for production while ensuring the code functions as expected.

Refer to this on the use of assertions:
https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/errorHandling.html#assertions

Jweng88 committed:
Improve code quality

There are various points in the code where code quality can be
improved, including complicated header comments and deep nesting.

Complicated header comments can be confusing and less clear.
Deep nesting can make it difficult for the reader
to keep track of the logic.

Complicated header comments are replaced with
simpler ones for better readability.
Deep nesting code are replaced with simpler code while
making the happy path prominent so that the happy path
is less nested, providing better readability and maintainability.

Simpler header comments are preferred for better readability.
Less nesting and making happy path prominent provides
better code readability and maintainability.

Refer to this on dealing with code quality:
https://nus-cs2103-ay2324s1.github.io/website/schedule/week5/topics.html#W5-4

Jweng88 committed:
Merge pull request #1 from Jweng88/branch-A-Assertions

Adding assertions to UI elements and task index

Jweng88 committed:
Merge branch 'master' into branch-A-CodeQuality

Jweng88 committed:
Merge pull request #2 from Jweng88/branch-A-CodeQuality

Improve Code quality

Jweng88 committed:
Add the ability to deal with duplicate items

When the user has duplicated tasks in the duke.txt file, or inputs
duplicated tasks, it will be added into the list of tasks.

This makes it confusing for the users when reading the list
of task as those duplicated tasks are redundant.

When loading the file and inputting new tasks, the code now checks
for duplicated tasks and only add the unique task.
In checking for duplicated tasks, the code compares the task
description, and the dateTime where applicable
without considering the seconds.

Checking for duplicated tasks is preferred for better clarity
and readability for users when viewing the list of tasks.
Ignoring the seconds when checking for duplicated tasks is preferred
as the seconds is insignificant when viewing tasks but provides
more flexibility in checking for duplicated task.

Refer to this to deal with duplicated tasks:
https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/projectDuke/index.html#extensions-category-c

Jweng88 committed:
Add a User Guide to the project

The code will now have a user guide for users
to understand how to use the application.

The lack of user guide would make it difficult for users
to know how to use the application.

Including the user guide allows users to have a
better idea on how to use the application.

A detailed user guide is preferred to ensure users understand
how to use the essential features of the application.

Jweng88 committed:
Add header comments

The code has a few header comments added.

The code had a few missing header comments
which might make it less readable.

Adding header comments make the ode more readable.

This is preferred to ensure better code quality and maintainability.
